Improve editor toolbar for `Control` nodes

YuriSizov opened the pull request:
Since https://github.com/godotengine/godot/pull/55630 is dead in waters for the time being, I've decided to extract the good parts from it, and specifically a better layout for anchor mode presets and container sizing options.

https://user-images.githubusercontent.com/11782833/180618055-ef87c30f-0284-4b89-a042-b3e126d55030.mp4

This also addresses concerns regarding changing anchor presets from the toolbar not changing grow direction and size flag options being behind that changes done in https://github.com/godotengine/godot/pull/55157.

-----

I've added a new control node, `PopupButton`, to serve as a base for `MenuButton` and an internal control that I've added for the toolbar. I needed to do this so I can share the switch on hover behavior, and use a different kind of popup at the same time. I figured, this was the best way. `MenuButton` should remain as is, functionality-wise.

__Closes https://github.com/godotengine/godot-proposals/issues/3559__

YuriSizov committed:
Improve editor toolbar for Control nodes

raulsntos commented on scene/gui/menu_button.cpp:
```diff
@@ -49,199 +109,80 @@ void MenuButton::shortcut_input(const Ref<InputEvent> &p_event) {
 }
 
 void MenuButton::_bind_methods() {
-	ClassDB::bind_method(D_METHOD("get_popup"), &MenuButton::get_popup);
```
I'm guessing we can't expose this method with a more specific return type in the derived class?

YuriSizov replied:
I don't think so.

YuriSizov commented on modules/mono/editor/GodotTools/GodotTools/Build/MSBuildPanel.cs:
```diff
@@ -135,11 +135,13 @@ public override void _Ready()
-            buildMenu.AddItem("Clean Solution".TTR(), (int)BuildMenuOptions.CleanSolution);
-            buildMenu.IdPressed += BuildMenuOptionPressed;
+            var buildMenu = (PopupMenu)_buildMenuBtn.GetPopup();
+            if (buildMenu != null) {
```
It shouldn't be null but since I have to cast here, I thought I'd add a check just in case.

raulsntos replied:
The cast will throw if the returned value is not cast-able to the requested type. Maybe you wanted to use `as` here? The `as` keyword will return null if it can't be cast-ed instead of throwing. But maybe I'm overthinking this, since we never call `SetPopup` it will never return anything other than the default which is the `PopupMenu` setup in the constructor of `MenuButton`.

YuriSizov replied:
I just mostly hoped it would work similar to our Object casting in C++, but I don't quite remember C#.

I can change it to whichever makes sense, just tell me what would be the best approach here.

raulsntos replied:
Since this will never fail, the cast is fine but the null check seems unnecessary and personally I think we want it throw if it fails rather than failing silently.

YuriSizov commented on doc/classes/PopupButton.xml:
```diff
@@ -0,0 +1,40 @@
+		<method name="get_popup" qualifiers="const">
+			<return type="Popup" />
+			<description>
+				Returns the [PopupMenu] contained in this button.
```
Yes, oops!

akien-mga commented:
What's the blank space?
![image](https://user-images.githubusercontent.com/4701338/181210383-44ebe6f5-617c-45c7-a8b2-ec2b31d7bb21.png)

YuriSizov commented:
> What's the blank space?

No idea, the file is [square](https://user-images.githubusercontent.com/11782833/181221984-f7bd8aa8-fbaa-468f-b422-f19362e591e0.png), and on GitHub it looks correct to me

![image](https://user-images.githubusercontent.com/11782833/181221749-c835bdbe-0f8a-47a4-95f3-1380454913d2.png)

akien-mga commented:
Weird, might be yet another GitHub regression on Firefox.

reduz commented on scene/gui/popup_button.h:
```diff
@@ -0,0 +1,63 @@
+#include "scene/gui/button.h"
+#include "scene/gui/popup.h"
+
+class PopupButton : public Button {
```
If you want to do this, then you should go all the way and handle OptionButton and ColorPicker, but I am not entirely sure this is justified.

YuriSizov replied:
What would you suggest? Users would expect the button to toggle on hover, just like MenuButton does. I could not do this, but then we'll likely get complaints that this behavior doesn't happen.

I don't think that OptionButton fits here, but ColorPickerButton for sure could be updated to use this new base node. But I'd prefer to do that in a separate PR, because this one doesn't depend on ColorPickerButton, and it doesn't touch what it doesn't need to touch, preserving compatibility.

KoBeWi commented on editor/plugins/control_editor_plugin.cpp:
```diff
@@ -448,37 +450,266 @@ bool EditorInspectorPluginControl::parse_property(Object *p_object, const Varian
+	switch (p_what) {
+		case NOTIFICATION_ENTER_TREE:
+		case NOTIFICATION_THEME_CHANGED: {
+			arrow_icon = get_theme_icon(SNAME("select_arrow"), SNAME("Tree"));
```
SNAME isn't needed here, because there are only 2 instances of this button.

YuriSizov replied:
Does it hurt though? I'd rather keep it so that the code is consistent, unless there is a reason not to do this.

KoBeWi replied:
At the definition of SNAME there is a comment that describes usage.
While abusing it *a bit* doesn't hurt, it shouldn't be used too much, because every SNAME uses up static memory.

YuriSizov replied:
Actually, this fetches tree's icon. So this is used beyond just this one file, so it's shared way more than just by two instances of this button, right?

KoBeWi replied:
Each SNAME uses unique memory, even if they contain the same string.

By "2 instances of this button" I meant that this particular line is executed only twice per theme change. SNAME is supposed to avoid StringName creation, because it's slow, but it only really matters for cases where it's called very often (e.g. all inspector controls should use SNAME, because they are instantiated with every inspector refresh, which affects responsiveness a lot). ControlEditorPopupButton is created only once per editor lifetime and only updates on theme change, which is rare, so it doesn't need SNAME.

YuriSizov replied:
Hmm, I see. I thought it optimized memory allocation throughout the entire codebase for similar strings. I didn't notice differences in usage of `SNAME` in the editor code though. And if the controls are coded correctly, with cache, this logic applies to all of them, so most theme fetching shouldn't really benefit from this.

Well, I can make the change either way, this is mostly my curiosity.

KoBeWi commented on editor/plugins/control_editor_plugin.cpp:
```diff
@@ -448,37 +450,266 @@ bool EditorInspectorPluginControl::parse_property(Object *p_object, const Varian
+			PopupPanel *pp = Object::cast_to<PopupPanel>(get_popup());
+			if (pp) {
+				pp->set_theme_type_variation("ControlEditorPopupButton");
+				//pp->add_theme_style_override();
```
Commented-out code.

KoBeWi commented on doc/classes/PopupButton.xml:
```diff
@@ -0,0 +1,40 @@
+<?xml version="1.0" encoding="UTF-8" ?>
+<class name="PopupButton" inherits="Button" version="4.0" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xsi:noNamespaceSchemaLocation="../class.xsd">
+	<brief_description>
+	</brief_description>
```
At least a brief description would be nice...

KoBeWi commented:
PopupButton crashes if there is no popup set.

Also, maybe the PopupButton change could be done in a separate PR? Then you could handle ColorPicker button too.
Although I guess it's not very useful by itself 🤔

KoBeWi commented on editor/plugins/control_editor_plugin.cpp:
```diff
@@ -448,37 +450,266 @@ bool EditorInspectorPluginControl::parse_property(Object *p_object, const Varian
+
+ControlEditorPopupButton::ControlEditorPopupButton() {
+	PopupPanel *popup_panel = memnew(PopupPanel);
+	popup_panel->hide();
```
It's hidden by default.

YuriSizov commented:
> PopupButton crashes if there is no popup set.

At any specific moment? Or just having it in a scene?

> Also, maybe the PopupButton change could be done in a separate PR? Then you could handle ColorPicker button too. Although I guess it's not very useful by itself 🤔

Indeed, it's unlikely to be very useful. To me it was a necessity to implement the activate-on-hover behavior without doing hacks and editor-specific code injection. I'm not 100% convinced this is the best solution, but I don't have other ideas. I could forgo the on-hover behavior entirely and we can revisit it if and when users complain. The introduction of the `PopupButton` as base type shouldn't be compat breaking even if we need to do this after 4.0.

KoBeWi commented:
> At any specific moment?

When it's pressed.

KoBeWi commented on editor/plugins/control_editor_plugin.cpp:
```diff
@@ -448,37 +450,266 @@ bool EditorInspectorPluginControl::parse_property(Object *p_object, const Varian
+	_add_row_button(extra_row, PRESET_FULL_RECT, TTR("Full Rect"));
+}
+
+void ContainerSizeFlagPicker::_add_row_button(HBoxContainer *p_row, const SizeFlags p_flag, const String &p_name) {
```
This duplicates ControlAnchorPresetPicker. I wonder if it's possible to merge these two into one class 🤔 (you could replace constructors with initialization methods)

YuriSizov replied:
Could make it into a helper method outside of the class. We don't use that pattern often, but sometimes we do.

KoBeWi replied:
No I commented here, but the classes have really much in common, including a constant that is defined twice (`grid_separation`). So maybe it's possible to merge them into one class and use setup methods for initialization instead of constructors. Much of their code wold be shared. I think the only problem would be methods that take flags as argument. You'd probably need to make them into ints.

KoBeWi commented:
What happened to "Anchors Only" option?

KoBeWi commented on editor/plugins/control_editor_plugin.cpp:
```diff
@@ -594,400 +822,208 @@ Vector2 ControlEditorToolbar::_position_to_anchor(const Control *p_control, Vect
 void ControlEditorToolbar::_notification(int p_what) {
 	switch (p_what) {
 		case NOTIFICATION_ENTER_TREE:
 		case EditorSettings::NOTIFICATION_EDITOR_SETTINGS_CHANGED: {
```
I just noticed now, but the EDITOR_SETTINGS_CHANGED is for handling theme update, no? Couldn't it be THEME_CHANGED?

KoBeWi commented on editor/plugins/control_editor_plugin.cpp:
```diff
@@ -594,400 +822,208 @@ Vector2 ControlEditorToolbar::_position_to_anchor(const Control *p_control, Vect
-
-	PopupMenu *p = anchor_presets_menu->get_popup();
-	p->connect("id_pressed", callable_mp(this, &ControlEditorToolbar::_popup_callback));
+	//
```
Empty comment.

YuriSizov commented:
> What happened to "Anchors Only" option?

Ever since the original drawer's work nobody came forward and explained how it can be useful. I removed it because it was complicating the UI at this point, and now that presets also affect the grow direction, would we need more flags?..

KoBeWi commented:
Well it's useful when you position a control, but forget to use anchors beforehand. It's rather rare, but I did use it a few times 🤔
(although the workaround is just copying the position value, so it's not much of a problem)

KoBeWi commented on scene/gui/popup_button.cpp:
```diff
@@ -0,0 +1,154 @@
+	ERR_FAIL_COND(!p_popup_node);
+	ERR_FAIL_COND_MSG(p_popup_node->get_parent() != this, "Connected popup must be a child of the button.");
+
+	if (popup) {
+		popup->disconnect("about_to_popup", callable_mp(this, &PopupButton::_popup_visibility_changed));
+		popup->disconnect("popup_hide", callable_mp(this, &PopupButton::_popup_visibility_changed));
+	}
```
I think replacing the popup is a bit awkward. You have to free it yourself, but you can't do it before setting the new popup. Maybe this should also automatically free the old popup?

YuriSizov commented:
> You mean switch on hover? It's only useful when two of the buttons are visible, which is rather uncommon when working with UI.

The way switch on hover is currently implemented (and this PR doesn't change) is less than ideal. It's parent based, and so it breaks in our editor UI when we talk about toolbars. Each toolbar has its own parent hbox, so only the buttons inside of the same toolbar can be activated on hover. As a user I'd expect the entire toolbar to be able to be hover-selected like that.

So it's not just the two buttons in this toolbar, but also the buttons from other toolbars and the main toolbar that should play into this. But, as mentioned, I didn't change that here. Having a common parent class like that would allow for future improvements in this area.

But, at this point, I'm tired of dragging these toolbar improvements for almost a year already. Every time they end up entangled with something else that prevents it from moving. So I'll amend this PR today to remove anything related to the `PopupButton`. And then maybe open a proposal for it to improve the switch on hover behavior.

YuriSizov commented:
Okay, removed any changes done to `MenuButton` and the concept of `PopupButton`. This resolves every related comment. I've addressed the rest of them.

For the picker popups I've created a new base class that pulled some of the shared code. But I didn't unite them entirely, because there is plenty of non-shared code in them too. But it should keep the code duplication to the minimum now.

YuriSizov commented:
Would appreciate an approval from @KoBeWi as he's been with me and this change through every iteration 🙃

YuriSizov commented:
Alright, seems very mutual :)
